sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -580,37 +599,59 @@ public void setup() throws Exception {
         preparedStatement = connection.prepareStatement(spec.getStatement());
       }
 
-      @StartBundle
-      public void startBundle() {
-        batchCount = 0;
-      }
-
       @ProcessElement
       public void processElement(ProcessContext context) throws Exception {
         T record = context.element();
 
-        preparedStatement.clearParameters();
-        spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
-        preparedStatement.addBatch();
-
-        batchCount++;
+        records.add(record);
 
-        if (batchCount >= spec.getBatchSize()) {
+        if (records.size() >= spec.getBatchSize()) {
           executeBatch();
         }
       }
 
+      private void processRecord(T record) {
+        try {
+          preparedStatement.clearParameters();
+          spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
+          preparedStatement.addBatch();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
       @FinishBundle
       public void finishBundle() throws Exception {
         executeBatch();
       }
 
-      private void executeBatch() throws SQLException {
-        if (batchCount > 0) {
-          preparedStatement.executeBatch();
-          connection.commit();
-          batchCount = 0;
+      private void executeBatch() throws SQLException, IOException, InterruptedException {
+        if (records.size() == 0) {
+          return;
+        }
+        Sleeper sleeper = Sleeper.DEFAULT;
+        BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
+        while (true) {
+          try {
+            // add each record in the statement batch
+            records.stream().forEach(this::processRecord);
+            // execute the batch
+            preparedStatement.executeBatch();
+            // commit the changes
+            connection.commit();
+            break;
+          } catch (SQLException exception) {
```

Comment:
It seems that all popular databases report deadlocks the same way - via an SQLState of `"40001"`.

https://db.apache.org/derby/docs/10.0/manuals/develop/develop75.html
https://docs.oracle.com/javadb/10.8.3.0/devguide/cdevconcepts32861.html
https://dev.mysql.com/doc/refman/5.7/en/innodb-deadlock-example.html
https://support.microsoft.com/en-us/help/963957/40001-microsoft-odbc-sql-server-driver-sql-server-your-transaction-was
https://stackoverflow.com/questions/26383624/postgres-exceptions-and-java

and so on.